src/box/lua/connpool.lua

```diff
@@ -271,46 +316,102 @@ local function filter(opts)
             table.insert(static_candidates, instance_name)
         end
     end
-    -- Return if retrieving dynamic information is not required.
-    if next(static_candidates) == nil or next(dynamic_opts) == nil then
+
+    if mode == 'any' then
         return static_candidates
     end
 
-    -- Filter the remaining candidates after connecting to them.
-    local connected_candidates = connect_to_candidates(static_candidates)
+    local connected_instances = acquire_all_candidates(static_candidates)
+
     local dynamic_candidates = {}
-    for _, instance_name in pairs(connected_candidates) do
+    for _, instance_name in pairs(connected_instances) do
         if is_candidate_match_dynamic(instance_name, dynamic_opts) then
             table.insert(dynamic_candidates, instance_name)
         end
     end
+
     return dynamic_candidates
 end
 
+-- Subroutine for the `call()` function. Tries to find an active
+-- connection matching the requirements and considering the
+-- specified priorities.
 local function get_connection(opts)
-    local mode = nil
+    assert(opts.mode == 'alive' or opts.mode == 'ro' or opts.mode == 'rw' or
+           opts.mode == 'prefer_ro' or opts.mode == 'prefer_rw')
+
+    local mode = 'alive'
     if opts.mode == 'ro' or opts.mode == 'rw' then
         mode = opts.mode
     end
-    local candidates_opts = {
+
+    -- It's better to use the local instance to perform the call
+    -- faster. So prefer the local instance by default.
+    local prefer_local = true
+    if opts.prefer_local ~= nil then
+        prefer_local = opts.prefer_local
+    end
+
+    local dynamic_opts = {
+        mode = mode,
+    }
+
+    local local_instance_name = box.info.name
+
+    local filter_static_opts = {
         groups = opts.groups,
         replicasets = opts.replicasets,
         instances = opts.instances,
         labels = opts.labels,
         roles = opts.roles,
         sharding_roles = opts.sharding_roles,
-        mode = mode,
     }
-    local candidates = filter(candidates_opts)
+    local static_candidates = filter(filter_static_opts)
+
+    -- In case we don't need to prioritize the matching instances
+    -- we can wait for any connection available.
+    if opts.mode ~= 'prefer_rw' and opts.mode ~= 'prefer_ro' then
+        if prefer_local then
+            for _, instance_name in ipairs(static_candidates) do
+                if instance_name ~= local_instance_name then
+                    goto continue
+                end
+
+                local conn = connect(local_instance_name,
+                                     { wait_connected = true })
+                assert(conn.state == 'active')
```

Comment:
TBH, I don't get why it have to succeed.

Reply:
Oh, didn't think of it. I wrapped it with a pcall and now check if everything is ok. Also, I've added a comment that the connect is still `wait_connected` since the call is almost always successfull.

Reply:
OK, this problem is resolved. I'm not sure that we should handle it this way, but I'll open a separate comment.

Reply:
Opened https://github.com/tarantool/tarantool/pull/10438#discussion_r1779330155 and https://github.com/tarantool/tarantool/pull/10438#discussion_r1779330694.